src/db/spanner/models.rs

```diff
@@ -0,0 +1,1302 @@
+        let user_id = params.user_id.legacy_id as u32;
+        let collection_id = self.get_or_create_collection_id(&params.collection)?;
+        if let Some(CollectionLock::Read) = self
+            .inner
+            .session
+            .borrow()
+            .coll_locks
+            .get(&(user_id, collection_id))
+        {
+            Err(DbError::internal("Can't escalate read-lock to write-lock"))?
+        }
+
+        // Lock the db
+        self.begin()?;
+        let spanner = &self.conn;
+        let session = spanner.session.name.as_ref().unwrap();
+        let mut sql = ExecuteSqlRequest::default();
+        sql.sql = Some("SELECT CURRENT_TIMESTAMP() as now, last_modified FROM user_collections WHERE userid=@userid AND collection=@collectionid FOR UPDATE".to_string());
+        let mut params = HashMap::new();
+        params.insert("userid".to_string(), user_id.to_string());
+        params.insert("collection".to_string(), collection_id.to_string());
+        sql.params = Some(params);
+
+        let results = spanner
+            .hub
+            .projects()
+            .instances_databases_sessions_execute_sql(sql, session)
+            .doit();
+        if let Ok(results) = results {
+            if let Some(rows) = results.1.rows {
+                let modified = SyncTimestamp::from_i64(rows[0][0].parse().unwrap())?;
+                self.session
+                    .borrow_mut()
+                    .coll_modified_cache
+                    .insert((user_id, collection_id), modified);
+            }
+        }
+        self.session
+            .borrow_mut()
+            .coll_locks
+            .insert((user_id, collection_id), CollectionLock::Write);
+
+        // let modified = user_collections::table
+        //     .select(user_collections::modified)
+        //     .filter(user_collections::user_id.eq(user_id as i32))
+        //     .filter(user_collections::collection_id.eq(collection_id))
+        //     .for_update()
+        //     .first(&self.conn)
+        //     .optional()?;
+        // if let Some(modified) = modified {
+        //     let modified = SyncTimestamp::from_i64(modified)?;
+        //     // Forbid the write if it would not properly incr the timestamp
+        //     if modified >= self.timestamp() {
+        //         Err(DbErrorKind::Conflict)?
+        //     }
+        //     self.session
+        //         .borrow_mut()
+        //         .coll_modified_cache
+        //         .insert((user_id, collection_id), modified);
+        // }
+        // self.session
+        //     .borrow_mut()
+        //     .coll_locks
+        //     .insert((user_id, collection_id), CollectionLock::Write);
+        Ok(())
+    }
+
+    pub(super) fn begin(&self) -> Result<()> {
+        let spanner = &self.conn;
+        let session = spanner.session.name.as_ref().unwrap();
+        let options = Some(TransactionOptions::default());
```

Comment:
TransactionOptions should differ depending on the call from lock_for_write/read (read/write vs plain read). Maybe begin should take a TransactionOptions argument